tools/bin/hawqregister

```diff
@@ -0,0 +1,272 @@
+        dburl = dbconn.DbURL(hostname=options.host, port=options.port, username=options.user, dbname=databasename)
+        conn = dbconn.connect(dburl, False)
+        rows = dbconn.execSQL(conn, query)
+	conn.commit()
+        if rows.rowcount == 0:
+            logger.error("target not found in table gp_distribution_policy")
+            sys.exit(1)
+        for row in rows:
+            if row[0] != None:
+                logger.error("Cannot register file(s) to a table which is hash-typed")
+                sys.exit(1)
+
+        conn.close()
+
+    except DatabaseError, ex:
+        logger.error("Failed to connect to database, this script can only be run when the database is up")
+        logger.error("host = %s, port = %d, user = %s, dbname = %s, query = %s" % (options.host, options.port, options.user, databasename, query))
+        sys.exit(1)
+
+
+def get_files_in_hdfs(filepath):
+    files = []
+    sizes = []
+    hdfscmd = "hadoop fs -test -e %s" % filepath
+    result = local_ssh(hdfscmd)
+    if result != 0:
+        logger.error("Path '%s' does not exist in hdfs" % filepath)
+        sys.exit(1)
+    
+    hdfscmd = "hadoop fs -ls -R %s" % filepath
+    result, out, err = local_ssh_output(hdfscmd)
+    outlines = out.splitlines()
+
+    # recursively search all the files under path 'filepath'
+    i = 0
+    for line in outlines:
+        lineargs = line.split()
+        if len(lineargs) == 8 and lineargs[0].find ("d") == -1:
+            files.append(lineargs[7])
+            sizes.append(int(lineargs[4]))
+
+    if len(files) == 0:
+        logger.error("Dir '%s' is empty" % filepath)
+        sys.exit(1)
+
+    return files, sizes
+
+
+def check_parquet_format(options, files):
+    # check whether the files are parquet format by checking the first and last four bytes
+    for file in files:
+        hdfscmd = "hadoop fs -cat %s | head -c 4 | grep PAR1" % file
+        result1 = local_ssh(hdfscmd)
+        hdfscmd = "hadoop fs -cat %s | tail -c 4 | grep PAR1" % file
+        result2 = local_ssh(hdfscmd)
+        if result1 or result2:
+            logger.error("File %s is not parquet format" % file)
+            sys.exit(1)
+
+
+def get_metadata_from_database(options, databasename, seg_name):
+    try:
+        query = "select segno from pg_aoseg.%s;" % seg_name
+        dburl = dbconn.DbURL(hostname=options.host, port=options.port, username=options.user, dbname=databasename)
+        conn = dbconn.connect(dburl, False)
+        rows = dbconn.execSQL(conn, query)
+	conn.commit()
+        conn.close()
+
+    except DatabaseError, ex:
+        logger.error("Failed to connect to database, this script can only be run when the database is up")
+        logger.error("host = %s, port = %d, user = %s, dbname = %s, query = %s" % (options.host, options.port, options.user, databasename, query))
+        sys.exit(1)
+
```

Comment:
> The main hawq register steps are as follows:
> get_files_in_hdfs(filepath)
> check_parquet_format(options, files)
> firstsegno = get_metadata_from_database(options, databasename, seg_name)
> move_files_in_hdfs(options, databasename, tablename, files, firstsegno, True)
> insert_metadata_into_database(options, databasename, tablename, seg_name, firstsegno, sizes)

If one of the reads fails, we just exit cause no file has been changed. If move_files_in_hdfs fails, the metadata will not be inserted. If insert_metadata_into_database fails, we will move the files back to their original location.